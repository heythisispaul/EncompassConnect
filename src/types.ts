/* eslint-disable camelcase */
/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

/**
 * @packageDocumentation
 * @module Interfaces
 */
export interface SortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: SortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: (PipeLineFilter | PipeLineTerms)[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectInitOptions {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
  version?: number;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
  version?: number;
  useTruncatedBase?: boolean;
}

export interface AssignMilestoneOptions {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestoneOptions {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean | string;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}

export interface BatchUpdateStatus {
  status: string;
  lastModified: string;
}

export interface BatchUpdate {
  getRequestId: () => string;
  getUpdateStatus: () => Promise<BatchUpdateStatus>;
}

export interface UpdateLoanWithGenerateContract {
  standardFields?: {
    [key: string]: any;
  },
  customFields?: {
    [key: string]: any;
  }
}

export interface TokenIntrospection {
  active: boolean;
  scope: string;
  client_id: string;
  username: string;
  token_type: 'Bearer';
  exp: number;
  sub: string;
  encompass_instance_id: string;
  user_name: string;
  user_key: string;
  encompass_user: string;
  identity_type: 'Enterprise';
  encompass_instance_type: string;
  encompass_client_id: string;
  realm_name: string;
  bearer_token?: string;
}

export interface ListOfUsersOptions {
  viewEmailSignature?: boolean;
  groupId?: string | number;
  roleId?: string;
  personaId?: string | number;
  organizationId?: string | number;
  userName?: string;
  start?: number;
  limit?: number;
}
