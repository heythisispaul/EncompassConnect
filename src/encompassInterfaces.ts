/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */
export interface sortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: sortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: PipeLineTerms[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectConstructor {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
}

export interface AssignMilestone {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestone {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}
