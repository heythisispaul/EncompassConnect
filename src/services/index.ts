export { default as LoanService } from './loans';
export { default as MilestoneService } from './milestones';
export { default as SchemaService } from './schemas';
export { default as UserService } from './users';
