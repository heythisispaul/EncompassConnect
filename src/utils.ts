import { FieldReaderResult } from './types';

export const massageCustomFields = (customFields: any) => {
  if (!customFields) {
    return [];
  }
  const massaged = Object
    .entries(customFields).map(([fieldName, value]: [string, any]) => {
      const parsedNumber = parseFloat(value);
      const customFieldObject: any = {
        fieldName,
        stringValue: value && value.toString ? value.toString() : value,
      };
      if (!Number.isNaN(parsedNumber)) {
        customFieldObject.numericValue = parsedNumber;
      }
      return customFieldObject;
    });
  return massaged;
};

export const objectToURLString = (input: any, queryString?: boolean): string => {
  // @ts-ignore
  const result = input ? new URLSearchParams(input).toString() : '';
  return queryString ? `?${result}` : result;
};

export const reduceFieldReaderValues = (fieldReaderResult: FieldReaderResult[]) => fieldReaderResult
  .reduce((acc, curr) => {
    const { fieldId, value } = curr;
    // @ts-ignore
    acc[fieldId] = value;
    return acc;
  }, {});
