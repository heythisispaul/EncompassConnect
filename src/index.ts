/* eslint-disable camelcase */
import fetch, {
  Response,
  RequestInit,
  HeadersInit,
} from 'node-fetch';
import {
  EncompassConnectConstructor,
  InternalRequestOptions,
  PipeLineContract,
  BatchLoanUpdateContract,
  AssignMilestone,
  UpdateMilestone,
  LoanUpdateOptions,
  batchUpdateStatus,
  UpdateLoanWithGenerateContract,
  BatchUpdate,
  TokenIntrospection,
} from './encompassInterfaces';
import { massageCustomFields } from './utils';

class EncompassConnect {
  #clientId: string;

  #APIsecret: string;

  #instanceId: string;

  #password: string;

  #token: string | null;

  username: string;

  base: string;

  authBase: string;

  version: number;

  constructor({
    clientId,
    APIsecret,
    instanceId,
    username,
    password,
    version = 1,
  }: EncompassConnectConstructor) {
    this.#clientId = clientId;
    this.#APIsecret = APIsecret;
    this.#instanceId = instanceId;
    this.#password = password || '';
    this.#token = '';
    this.username = username || '';
    this.version = version;
    this.base = 'https://api.elliemae.com/encompass';
    this.authBase = 'https://api.elliemae.com';
  }

  setToken(token:string | null): void {
    this.#token = token;
  }

  private withTokenHeader(headers = {}): HeadersInit {
    return {
      ...headers,
      Authorization: `Bearer ${this.#token}`,
    };
  }

  private async fetchWithRetry(
    path: string,
    options: RequestInit = {},
    customOptions: InternalRequestOptions = {},
  ): Promise<any> {
    const { isRetry, isNotJson, version } = customOptions;
    const shouldRetry = !isRetry && this.username && this.#password;
    const failedAuthError = new Error(
      `Token invalid. ${!isRetry ? 'Will reattempt with new token' : 'Unable to get updated one.'}`,
    );
    try {
      if (!this.#token) {
        await this.getToken();
      }
      const url = `${this.base}/v${version || this.version}${path}`;
      const optionsWithToken: RequestInit = {
        ...options,
        headers: this.withTokenHeader(options.headers),
      };
      const response: Response = await fetch(url, optionsWithToken);
      if (response.status === 401) {
        throw failedAuthError;
      }
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return isNotJson ? response : await response.json();
    } catch (error) {
      if (shouldRetry && error === failedAuthError) {
        this.setToken(null);
        return this.fetchWithRetry(path, options, {
          ...customOptions,
          isRetry: true,
        });
      }
      throw error;
    }
  }

  async getToken(username?: string, password?: string): Promise<void> {
    // @ts-ignore
    const body: string = new URLSearchParams({
      grant_type: 'password',
      username: `${username || this.username}@encompass:${this.#instanceId}`,
      password: password || this.#password,
      client_id: this.#clientId,
      client_secret: this.#APIsecret,
    }).toString();

    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body,
      redirect: 'follow',
    };

    const response = await fetch(`${this.authBase}/oauth2/v1/token`, requestOptions);
    const { access_token } = await response.json();
    this.setToken(access_token);
  }

  async introspectToken(token?: string): Promise<TokenIntrospection | null> {
    // @ts-ignore
    const body: string = new URLSearchParams({
      token: token || this.#token,
    }).toString();
    const Authorization: string = `Basic ${Buffer.from(`${this.#clientId}:${this.#APIsecret}`).toString('base64')}`;
    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        Authorization,
      },
      body,
    };
    try {
      const response = await fetch(`${this.authBase}/oauth2/v1/token/introspection`, requestOptions);
      if (response.ok) {
        const tokenData: TokenIntrospection = await response.json();
        return tokenData;
      }
      return null;
    } catch (error) {
      return null;
    }
  }

  async getCanonicalNames(): Promise<any> {
    const canonicalNames = await this.fetchWithRetry('/loanPipeline/fieldDefinitions');
    return canonicalNames;
  }

  async viewPipeline(options: PipeLineContract, limit?: number) {
    const uri = `/loanPipeline${limit ? `?limit=${limit}` : ''}`;
    const pipeLineData = await this.fetchWithRetry(uri, {
      method: 'POST',
      headers: this.withTokenHeader(),
      body: JSON.stringify(options),
    });
    return pipeLineData;
  }

  async batchLoanUpdate(options: BatchLoanUpdateContract): Promise<BatchUpdate> {
    const response: any = await this.fetchWithRetry('/loanBatch/updateRequests', {
      method: 'POST',
      body: JSON.stringify(options),
    }, { isNotJson: true });
    const requestId: string = response && response.headers
      ? response.headers.get('location').split('/').reverse()[0]
      : null;
    return {
      getRequestId: () => requestId,
      getUpdateStatus: async () => {
        const url = `/loanBatch/updateRequests/${requestId}`;
        const status: batchUpdateStatus = await this.fetchWithRetry(url);
        return status;
      },
    };
  }

  async request(url: string, options: RequestInit): Promise<Response> {
    const response = await this.fetchWithRetry(url, options, { isNotJson: true });
    return response;
  }

  schemas = {
    generateContract: async (fields: any): Promise<any> => {
      const contract = await this.fetchWithRetry(
        '/schema/loan/contractGenerator',
        { method: 'POST', body: JSON.stringify(fields) },
      );
      return contract;
    },
    getLoanSchema: async (entities?: string[]): Promise<any> => {
      const entitiesString = entities ? entities.toString() : '';
      const url = `/schema/loan${entitiesString ? `?entities=${entitiesString}` : ''}`;
      const response = await this.fetchWithRetry(url);
      return response;
    },
  }

  loans = {
    getGuidByLoanNumber: async (loanNumber: string): Promise<string> => {
      const [loanResult] = await this.viewPipeline({
        filter: {
          operator: 'and',
          terms: [
            {
              canonicalName: 'Loan.LoanNumber',
              matchType: 'exact',
              value: loanNumber,
            },
          ],
        },
      });
      return loanResult ? loanResult.loanGuid : null;
    },
    get: async (guid: string, entities?: string[]): Promise<any> => {
      const url = `/loans/${guid}${entities ? `?entities=${entities.toString()}` : ''}`;
      const data: any = await this.fetchWithRetry(url);
      return data;
    },
    update: async (guid: string, loanData: any, options?: LoanUpdateOptions): Promise<any> => {
      const defaultOptions = [
        ['appendData', 'false'],
        ['persistent', 'transient'],
        ['view', 'entity'],
      ];
      // @ts-ignore
      const queryOptions = new URLSearchParams(options || defaultOptions).toString();
      const url = `/loans/${guid}?${queryOptions}`;
      const fetchOptions: RequestInit = {
        method: 'PATCH',
        body: JSON.stringify(loanData),
      };
      await this.fetchWithRetry(url, fetchOptions, { isNotJson: true });
    },
    updateWithGeneratedContract: async (
      guid: string,
      loanData: UpdateLoanWithGenerateContract,
      options?: LoanUpdateOptions,
    ): Promise<void> => {
      const { customFields, standardFields } = loanData;
      const massagedCustomFields = massageCustomFields(customFields);
      const generatedContract: Promise<any> = await this.schemas.generateContract(standardFields);
      const generatedLoanData = { ...generatedContract, customFields: massagedCustomFields };
      await this.loans.update(guid, generatedLoanData, options);
    },
    delete: async (guid: string): Promise<void> => {
      await this.fetchWithRetry(
        `/loans/${guid}`,
        { method: 'DELETE' },
        { isNotJson: true },
      );
    },
  }

  milestones = {
    get: async (guid: string) => {
      const milestones: any[] = await this.fetchWithRetry(`/loans/${guid}/milestones`);
      return milestones;
    },
    assign: async ({
      milestone,
      userId,
      loanGuid,
    }: AssignMilestone): Promise<void> => {
      const milestoneData = await this.milestones.get(loanGuid);
      const matchingMilestone = milestoneData
        .find(({ milestoneName }) => milestone === milestoneName);
      if (!matchingMilestone) {
        throw new Error(`No milestone found for loan ${loanGuid} matching name "${milestone}"`);
      }
      const { id } = matchingMilestone;
      const fetchOptions = {
        method: 'PUT',
        body: JSON.stringify({
          loanAssociateType: 'User',
          id: userId,
        }),
      };
      await this.fetchWithRetry(`/loans/${loanGuid}/associates/${id}`, fetchOptions, { isNotJson: true });
    },
    update: async ({
      loanGuid,
      milestone,
      options,
      action,
    }: UpdateMilestone): Promise<void> => {
      const milestoneData = await this.milestones.get(loanGuid);
      const matchingMilestone = milestoneData
        .find(({ milestoneName }) => milestone === milestoneName);
      if (!matchingMilestone) {
        throw new Error(`No milestone found for loan ${loanGuid} matching name "${milestone}"`);
      }
      const { id } = matchingMilestone;
      const fetchOptions = {
        method: 'PATCH',
        body: JSON.stringify(options),
      };
      const uri = `/loans/${loanGuid}/milestones/${id}${action ? `?action=${action}` : ''}`;
      await this.fetchWithRetry(uri, fetchOptions, { isNotJson: true });
    },
  }
}

export default EncompassConnect;
