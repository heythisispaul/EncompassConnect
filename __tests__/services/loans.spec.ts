import querystring from 'querystring';
import { testInstanceWithToken, mockGuid } from '../__utils__/mockEncompassConnectInstances';
import { mockResponse, fetch } from '../__utils__/mockFunctions';
import { matchesFetchSnapshot, returnsResponseBody } from '../__utils__/commonTests';
import { LoanUpdateOptions, UpdateLoanWithGenerateContract } from '../../src/types';

jest.mock('node-fetch');

describe('loans', () => {
  const testLoanUpdateData = {
    applications: [
      {
        id: 'borrower_1',
        borrower: {
          firstName: 'Todd',
          lastName: 'Chavez',
        },
      },
    ],
  };

  beforeEach(() => {
    fetch.mockReset();
  });

  describe('getGuidByLoanNumber', () => {
    const testLoanNumber = '1234567';

    it('returns null if there is no matching loan found', async () => {
      mockResponse([]);
      const guid = await testInstanceWithToken.loans.getGuidByLoanNumber(testLoanNumber);
      expect(guid).toEqual(null);
    });

    it('returns the loanGuid value of the first index of the pipeline response', async () => {
      const mockLoanGuid = 'loan guid value';
      mockResponse([{ loanGuid: mockLoanGuid }]);
      const guid = await testInstanceWithToken.loans.getGuidByLoanNumber(testLoanNumber);
      expect(guid).toEqual(mockLoanGuid);
    });
  });

  describe('get', () => {
    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('loans.get', mockGuid)
    ));

    it('returns the body of the response', async () => (
      returnsResponseBody('loans.get', mockGuid)
    ));

    it('adds the entities from the provided array to the query string', async () => {
      const entities = ['diane', 'nguyen'];
      mockResponse();
      await testInstanceWithToken.loans.get(mockGuid, entities);
      // @ts-ignore
      const path: string = fetch.mock.calls[0][0];
      expect(path.includes('?entities=diane,nguyen')).toBeTruthy();
    });
  });

  describe('update', () => {
    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('loans.update', mockGuid, testLoanUpdateData)
    ));

    it('uses the optional options if provided', async () => {
      mockResponse();
      const options: LoanUpdateOptions = {
        appendData: 'true',
        persistent: 'permanent',
        view: 'id',
      };
      await testInstanceWithToken.loans.update(mockGuid, testLoanUpdateData, options);
      // @ts-ignore
      const url: string = fetch.mock.calls[0][0];
      const usedQueryString = querystring.parse(url.split('?')[1]);
      expect(usedQueryString).toEqual(options);
    });
  });

  describe('updateWithGeneratedContract', () => {
    it('makes the API call with the correct information', async () => {
      mockResponse({
        testContract: '<GENERATE CONTRACT RESPONSE>',
      });
      mockResponse();
      const testData: UpdateLoanWithGenerateContract = {
        standardFields: {
          317: 'Bojack Horseman',
        },
        customFields: {
          'CX.TEST.FIELD': 'test value',
        },
      };
      await testInstanceWithToken.loans.updateWithGeneratedContract(mockGuid, testData);
      expect(fetch.mock.calls[1]).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('loans.delete', mockGuid)
    ));
  });

  describe('fieldReader', () => {
    const fields = ['4000', '4002'];
    const fieldReaderResponse = [
      { fieldId: 'field1', value: 'value 1' },
      { fieldId: 'field2', value: 'value 2' },
    ];

    it('makes the API call with the correct information', async () => (
      matchesFetchSnapshot('loans.fieldReader', mockGuid, fields)
    ));

    it('appends the query string if the includeMetadata flag is set to true', async () => (
      matchesFetchSnapshot('loans.fieldReader', mockGuid, fields, { includeMetadata: true })
    ));

    it(' calls the reduceFieldReaderValues side effect if the mapResponse key is true', async () => {
      mockResponse(fieldReaderResponse);
      const result = await testInstanceWithToken.loans
        .fieldReader(mockGuid, fields, { mapResponse: true });
      expect(result).toEqual({ field1: 'value 1', field2: 'value 2' });
    });
  });
});
