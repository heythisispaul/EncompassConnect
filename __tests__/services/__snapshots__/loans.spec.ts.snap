// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loans delete makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "DELETE",
  },
]
`;

exports[`loans fieldReader appends the query string if the includeMetadata flag is set to true 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>/fieldReader?includeMetadata=true",
  Object {
    "body": "[\\"4000\\",\\"4002\\"]",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "POST",
  },
]
`;

exports[`loans fieldReader makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>/fieldReader",
  Object {
    "body": "[\\"4000\\",\\"4002\\"]",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "POST",
  },
]
`;

exports[`loans get makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`loans update makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>?appendData=false&persistent=transient&view=entity",
  Object {
    "body": "{\\"applications\\":[{\\"id\\":\\"borrower_1\\",\\"borrower\\":{\\"firstName\\":\\"Todd\\",\\"lastName\\":\\"Chavez\\"}}]}",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "PATCH",
  },
]
`;

exports[`loans updateWithGeneratedContract makes the API call with the correct information 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/loans/<MOCK GUID VALUE>?appendData=false&persistent=transient&view=entity",
  Object {
    "body": "{\\"testContract\\":\\"<GENERATE CONTRACT RESPONSE>\\",\\"customFields\\":[{\\"fieldName\\":\\"CX.TEST.FIELD\\",\\"stringValue\\":\\"test value\\"}]}",
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
    "method": "PATCH",
  },
]
`;
